File: joss/comparisons/cat-2-type-safety/25-column-type-constraint/example-column-type-constraint.py
"""
Error Class 25: Column Type Constraint in Specialized Verbs

Python lets you apply numeric functions to any column via
select_dtypes or apply. If the column isn't actually numeric,
the operation fails at runtime or produces wrong results.
"""
import pandas as pd
import numpy as np

patients = pd.DataFrame({
    "name": ["Alice", "Bob"],
    "age": [30, 45],
    "weight": [65.5, 80.0],
    "insurance": ["Medicare", "Medicaid"],
})

# Runtime error: log on string column
# np.log(patients[["age", "insurance"]])  # TypeError at runtime

# SILENT: select_dtypes("number") works, but if you manually select wrong columns...
# No compile-time check that your column list matches the expected type
result = patients[["age", "insurance"]].apply(lambda x: x * 2)
# age gets doubled (60, 90)
# insurance gets string-repeated: "MedicareMedicare", "MedicaidMedicaid"
print(result)

File: joss/comparisons/cat-2-type-safety/25-column-type-constraint/example-column-type-constraint.R
# Error Class 25: Column Type Constraint in Specialized Verbs
#
# R's across(where(is.numeric), fn) correctly filters columns at runtime,
# but if you manually specify columns, there's no compile-time check
# that they match the expected type.

library(tidyverse)

patients <- tibble(
  name = c("Alice", "Bob"),
  age = c(30, 45),
  weight = c(65.5, 80.0),
  insurance = c("Medicare", "Medicaid")
)

# across(where(is.numeric)) correctly selects numeric columns at runtime
result <- patients %>% mutate(across(where(is.numeric), log))
print(result)

# But manually specifying wrong columns errors at runtime:
# patients %>% mutate(across(c(age, insurance), log))
# Error: non-numeric argument to mathematical function
